### PR TITLE
Tcp streams with hektor-fsm

### PR DESCRIPTION
**Pull request by Sebitosh**
Hello,

This is my work in progress for aboutsip/pkts#147 , I am making a draft PR because the additions are starting to become a bit much (and because I was quite slow at making them).

At first I was going for a "standard OOP" approach for the Tcp stream logic, but since the comment on the issue I looked at the hektor library and now I am trying to use it to handle the stream logic. The biggest steps now are linking the FSM to the TcpStreamHandler and of course making tests for all of this. I also looked at the snice ConnectionId, but it looked a bit complicated for the usage I have here, so I worked out a TransportStreamId. This still has my attention for the coming days (as I hope to finish this work in not so long), and any feedback is appreciated (at this stage especially on the FSM, I need to document it of course but here I am kind of hoping I understood and used the hektor-fsm library correctly). I also spend a lot of time understanding the libraries and defining the streams exactly. I should be faster at producing code from now on.

From here I still need to:
- [x] Make tests for the FSM
- [x] Refactor some of my initial work around the use of the tcp FSM
- [x] Make the streamhandler use the TcpStreamFSM for it's processFrame function
- [x] Make tests for the different data structures (transportstreamid, tcpstream, ...)
- [x] Document code (except duplicate and stream handler)
- [x] Cleanup code (except duplicate and stream handler)
- [x] Make end to end tests interpreting pcap files and cheking if they correctly organize the TCP streams for general cases
- [x] Make end to end tests for corner cases with synthetic pcaps where the 5-tuple of a stream is reused for multiple connections
- [x] Correct and Define the logic for the handling of the corner cases
- [x] Make example file
- [x] Last clean/doc before review

### CODE REVIEW

File: pkts-streams/src/main/java/io/pkts/streams/impl/tcpFSM/TcpStreamFSM.java [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        final StateBuilder<TcpState, TcpStreamContext, TcpStreamData> closing_1_closing_2 = builder.withState(CLOSING_1_CLOSING_2);
+        final StateBuilder<TcpState, TcpStreamContext, TcpStreamData> closed_1_closing_2 = builder.withState(CLOSED_1_CLOSING_2);
+        final StateBuilder<TcpState, TcpStreamContext, TcpStreamData> closing_1_closed_2 = builder.withState(CLOSING_1_CLOSED_2);
+        final StateBuilder<TcpState, TcpStreamContext, TcpStreamData> closed = builder.withFinalState(CLOSED);
+
+        // define all transitions
+        init.transitionTo(HANDSHAKE).onEvent(TCPPacket.class).withGuard(TcpStreamFSM::isSynPacket);
+        init.transitionTo(FIN_WAIT_1).onEvent(TCPPacket.class).withGuard(TcpStreamFSM::isFinPacket);
+        init.transitionTo(CLOSED).onEvent(TCPPacket.class).withGuard(TcpStreamFSM::isRstPacket);
+        init.transitionTo(ESTABLISHED).asDefaultTransition();
```
Comment (jonbo372) [inline pkts-streams/src/main/java/io/pkts/streams/impl/tcpFSM/TcpStreamFSM.java:42]: You should avoid default transitions unless you have transient states, which you don't. For a fully functioning FSM, it should be "easy" to determine it's correctness by specifying all valid transitions. Any invalid transitions should error out.

So, I would recommend avoiding the default transition and ensure you specify all cases where it is legal to transition to, in this case, ESTABLISHED from INIT.

File: pkts-streams/src/main/java/io/pkts/streams/impl/tcpFSM/TcpStreamFSM.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+public class TcpStreamFSM{
+
+    public enum TcpState{
+        INIT, HANDSHAKE, ESTABLISHED, FIN_WAIT_1, FIN_WAIT_2,
+        CLOSING_1_CLOSING_2, CLOSED_1_CLOSING_2, CLOSING_1_CLOSED_2, CLOSED
+    }
+
+    public final static Definition<TcpState, TcpStreamContext, TcpStreamData> definition;
+
+    static {
```
Comment (jonbo372) [inline pkts-streams/src/main/java/io/pkts/streams/impl/tcpFSM/TcpStreamFSM.java:21]: A bit of a nit; but copy/paste the TCP state machine into the top here as it is much easier to understand what's going on.

**Review by jonbo372 — COMMENTED** [review]
Hey, sorry for late review. I just focused on the FSM and overall, seems that you're getting the idea of it. I just had a few comments. Looking forward to the "full" pull request, at which point I will look through it all in more detail.

**Comment by Sebitosh** [comment]
> Hey, sorry for late review. I just focused on the FSM and overall, seems that you're getting the idea of it. I just had a few comments. Looking forward to the "full" pull request, at which point I will look through it all in more detail.

Thank you for the feedback ! I got a bit delayed but I am working on it this week and next week, I hope until completion.

**Comment by Sebitosh** [comment]
I am currently in the process of trying out classifying different TCP streams. I noticed an interesting difference with how Wireshark handles streams compared to my code: when comparing the classifying from a complex pcap file, my code has may have more streams than Wireshark. 

This is due to how RST packets are handled: I consider the presence of a RST packet as a closing of the stream, considering that any packet with the same identifier must belong to a whole new stream a never to a closed stream. Wireshark on the other hand considers that packets from the same stream may arrive after RST packets.

I am unsure on how Wireshark computes that, I would guess by keeping track of both last arrived sequence numbers of the stream and never considering that a stream can not receive any new packets anymore (which I do, I consider that "closed" streams do not receive packets anymore). Should I try to stick as close as possible to how the Wireshark classifier behaves ?

I'll leave this question for now and focus on unit testing.

**Comment by Sebitosh** [comment]
> Edit: the hole in my reasonning here is that TCP packets with special flags could be duplictates or retransmission, and that using the sequence numbers is a good way of knowing if it is a duplicate (the chance of the sequence numbers colluding between streams like this is very small). I am reading up some comments in the wireshark code to understand their logic and then implement it here.

I still have some testing to do has I am still miss placing some packets in un-necessary new streams is some case(s) I still have not identified, but I am getting there with the last 2 commits

**Comment by jonbo372** [comment]
Thank you for all your work on this so far! It looks to be coming along nicely! 

And yeah, figuring out all the various cases with re-transmissions and how to deal with it can be quite a lot of work. I built a recording app for audio streams coming in over RTP once and it probably took a few years of heavy use in production before I felt I had found "all" cases and closed the loop on those. Reading wireshark code is probably a good idea as it has had a lot of eyes on it over the years.

I just had a quick look over your code but will refrain from a full code review until you feel that you are 100% ready for one.

Btw, the pcaps you've checked in, just make sure they don't contain any sensitive information (i haven't look so perhaps you've already made sure of this).

Again, thank you!

**Comment by Sebitosh** [comment]
> Thank you for all your work on this so far! It looks to be coming along nicely!
> 
> And yeah, figuring out all the various cases with re-transmissions and how to deal with it can be quite a lot of work. I built a recording app for audio streams coming in over RTP once and it probably took a few years of heavy use in production before I felt I had found "all" cases and closed the loop on those. Reading wireshark code is probably a good idea as it has had a lot of eyes on it over the years.
> 
> I just had a quick look over your code but will refrain from a full code review until you feel that you are 100% ready for one.
> 
> Btw, the pcaps you've checked in, just make sure they don't contain any sensitive information (i haven't look so perhaps you've already made sure of this).
> 
> Again, thank you!

For now I still have issues with TCP Keep Alives beeing exchanged after a connection closes and their is still some issue with how I keep track of the different seq/ack numbers, so I still have some debugging to do.

The pcaps should not contain sensitive info. Most of them are synthetically generated, the biggest one is from some basic web browsing to known websites where I did not login to anything in particular.

I am a bit busy at the moment (lots of university projects to turn in, and I've got some finals to work on after that), but as soon as I can I will continue to find what is wrong and clean up the code.

**Comment by Sebitosh** [comment]
Hello, after my exams and some health issues, I am back working on this PR again !

Here I pushed my first full attempt at handling duplicates and other cases of packets belonging to already closed streams. It classifies packets in the exact same streams as wireshark does for my example capture with 272 streams. However, the logic to handle those cases is certainly not complete yet. I will return to it later to focus first on making the rest of the code ready for review.

In the coming days I will focus on cleaning and documenting the code, and I will produce some small tests for the new objects along with an example file to show how to use the streams. After that I will inspect further the corner cases of the duplicate handler.

The logic I was working on is in tcpDuplicateHandler. It takes care of the following cases were a packet belongs to a stream that seems to have already been closed:
 * 1. If the received packet corresponds to a duplicate of one of the signaling packets closing the connection (RST, FIN, or ACK of FIN)
 * 2. If the received packet corresponds to a packet that is out of order but still within the expected sequence numbers
 * 3. If the received packet corresponds to a packet that is in sequence with the last packet of the stream, which happens when one or both of the TCP clients did not understand the connection is closed (and this is were things can get _weird_ )

**Comment by jonbo372** [comment]
Hey! I see the pull requests coming through, very exciting and again, thanks for all your hard work.

I just wanted to let you know that I'm currently out traveling so will not get to this until second week in August, in case you're wondering why you don't hear anything from me.

Looking forward to it!

**Comment by Sebitosh** [comment]
> Hey! I see the pull requests coming through, very exciting and again, thanks for all your hard work.
> 
> I just wanted to let you know that I'm currently out traveling so will not get to this until second week in August, in case you're wondering why you don't hear anything from me.
> 
> Looking forward to it!

No worries, have a good vacation ! 

I'll finish up today (I only have some cleanup left to do before it's ready for review). I will also write a post with my conclusions on the corner cases to explain the behavior of the handler.

**Comment by Sebitosh** [comment]
To sum up how I handle the special cases of a connection having it's 5-tuple reused in the same file:

After some investigation, Wireshark does not use sequence numbers analysis nor does it use timestamps to classify connections using the same 5-tuple into different streams. In fact, **it only splits a stream with the same 5-tuple in one case: when a new syn packet has been identified** on the stream. In that case, it will consider any new packets to belong to that new stream, whatever the flags or sequence numbers. So, instead of continuing on with seq/ack analysis, I rolled back and modified the FSM to reflect the fact that the terminal state of a TCP connection for the handler is when the 5-tuple is reused by a new syn packet. This eliminates the problem of finding out what packets to attribute packets to a closed connection (which is great considering that in captured traffic their is actually a lot of _weird_ situations where tcp clients will exchange packets even when the connection is expected to be closed). 

It does however raise the question of when should the handler call the endStream function of the streamListener. For now, it calls it if after adding a packet if the state is closed or closed_ports_reused, meaning it will be called each time a packet is added to a closed connection. This contradicts the comment I found of the endStream function that says the function should be called only once to not create confusions when writing out the stream. However, the "easy" alternatives to this are calling endStream only when ports are reused or only once when the stream reaches the closed state and never after despite adding packets to it. The writing problem the endStream documentation mentions should not be a problem here as writing out a DefaultTcpStream is still unsupported though.

I left out implementing a timeout mechanism as Wireshark does not split streams like that (even if packets have arrival times years apart) and because I am unsure on how to do it. For the endStream call I did not find how to call it when the file is over, and I am unsure if it is necessary to do so. Important to mention that the problem with calling endStream multiple times concerns only a minority of streams I saw in my captures. I generated synthetic traffic with all the corner cases I examined and used them in unit tests to see if the handler would classify them in the same way as Wireshark (when Wireshark splits a stream because of a new syn packet, it will mark that new syn with [Ports Reused]). Important also is that MPTCP traffic will not be recognized and put together.

With all this said, I believe my work to be ready for review. I'll be there in August to incorporate all the feedback !

**Comment by Sebitosh** [comment]
Hey @jonbo372  ! I hope you are doing well.

Do you have time to review this any time soon ? (No worries if you don't, I should be able to make time when you get to it eventually).

I remain at your disposal for any questions or changes :)

**Comment by kingthorin** [comment]
@jonbo372 anything we can do to help this along?

**Comment by jonbo372** [comment]
Hey @kingthorin and apologizes, this had completely slipped my mind! I will get to it this weekend and thank you for being so patient.

**Comment by Sebitosh** [comment]
I should be available next week to incorporate any feedback!

File: pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    /**
+    * Class implementing a {@link TransitionListener} to check if a stream closes due to a new SYN packet
+    *  in which case a new stream has to be started from that same closing SYN packet.
+    * */
+    private class SynListener implements TransitionListener<TcpState>{
+        @Override
+        public void onTransition(TcpState currentState, TcpState toState, Object o) {
+            if (!(o instanceof TCPPacket)){
+                throw new RuntimeException("Object passed to TcpFSM is of Class: " + o.getClass());
+            } else if (toState == TcpState.CLOSED_PORTS_REUSED && ((TCPPacket) o).isSYN()) {
```
Comment (jonbo372) [inline pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java:61]: no need for `else if` here as you are already bailing out.

File: pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+    /**
+    * Class implementing a {@link TransitionListener} to check if a stream closes due to a new SYN packet
+    *  in which case a new stream has to be started from that same closing SYN packet.
+    * */
+    private class SynListener implements TransitionListener<TcpState>{
+        @Override
+        public void onTransition(TcpState currentState, TcpState toState, Object o) {
+            if (!(o instanceof TCPPacket)){
+                throw new RuntimeException("Object passed to TcpFSM is of Class: " + o.getClass());
```
Comment (jonbo372) [inline pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java:60]: hmmm, isn't Hektor already guarding against this? And as you're checking this on the onTransition, isn't it a bit too late to complain about it? (man, I don't even recall how my own FSM framework is working! Guess I got to go back and read through it again :-)

File: pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+                final boolean isAlreadyClosed = stream.Ended();
+                stream.addPacket(tcpPacket);
+                this.notifyPacketReceived(stream, tcpPacket);
+                if (!isAlreadyClosed && stream.Ended()){ // call endStream only once when the last packet closed it
+                    this.notifyEndStream(stream);
+                }
+            }
+
+        } catch (Exception e){
+            e.printStackTrace();
```
Comment (jonbo372) [inline pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java:194]: should it be handled in some way?

File: pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+            stream = (stream == null) ? activeTcpStreams.get(pktStreamId.oppositeFlowDirection()) : stream;
+
+            if (stream == null) {
+                startNewStream(tcpPacket);
+            } else {
+                final boolean isAlreadyClosed = stream.Ended();
+                stream.addPacket(tcpPacket);
+                this.notifyPacketReceived(stream, tcpPacket);
+                if (!isAlreadyClosed && stream.Ended()){ // call endStream only once when the last packet closed it
+                    this.notifyEndStream(stream);
```
Comment (jonbo372) [inline pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java:189]: this is an "external" call right? If so, we should guard it with a catch-all as we do not want "customer code" to be able to interrupt our execution.

File: pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+    }
+
+    private IPPacket handleFragmentation(final IPPacket ipPacket) {
+        if (this.fragmentListener == null) {
+            return null;
+        }
+        try {
+            return this.fragmentListener.handleFragment(ipPacket);
+        } catch (final Throwable t) {
+            logger.warn("Exception thrown by FragmentListener when processing the IP frame", t);
```
Comment (jonbo372) [inline pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java:205]: As i often ask my team, if you log on warn, what do you want me to do at 3am when the pager goes off? So, what do you want me to do? As this is "customer" driven, I wouldn't want to log on warn (unless you do have a reason to). Just log on debug here. For real-world production code, I would want to have an entire sub-system where customer driven errors are bubbled up and reported to them. This is not it but hopefully you see my point

File: pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+    private void notifyPacketReceived(final TcpStream stream, final TCPPacket pkt) {
+        if (this.tcpListener != null) {
+            this.tcpListener.packetReceived(stream, pkt);
+        }
+    }
+
+    private void notifyEndStream(final TcpStream stream) {
+        if (this.tcpListener != null) {
+            this.tcpListener.endStream(stream);
+        }
```
Comment (jonbo372) [inline pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java:225]: a technique I often deploy to avoid having to do null-checks all the time is just to have a dummy, in this case, TcpListener that'll do nothing. That way, I do not have to check for null as my dummy would just be a NOOP...

File: pkts-streams/src/main/java/io/pkts/streams/impl/TransportStreamId.java [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+    public TransportStreamId(TCPPacket packet){
+        this.sourceAddress = packet.getParentPacket().getSourceIP();
+        this.destinationAddress = packet.getParentPacket().getDestinationIP();
+        this.sourcePort = packet.getSourcePort();
+        this.destinationPort = packet.getDestinationPort();
+        this.protocol = packet.getProtocol();
+    }
+
+    public TransportStreamId oppositeFlowDirection(){
```
Comment (jonbo372) [inline pkts-streams/src/main/java/io/pkts/streams/impl/TransportStreamId.java:42]: nit; and naming is hard but perhaps `remoteTransportStreamId`?

File: pkts-streams/src/main/java/io/pkts/streams/impl/tcpFSM/TcpStreamFSM.java (hunk 2 above)
Comment (jonbo372) [reply pkts-streams/src/main/java/io/pkts/streams/impl/tcpFSM/TcpStreamFSM.java:27]: Still waiting for that one :-)

**Review by jonbo372 — CHANGES_REQUESTED** [review]
Good stuff! Just some minor comments and then, please squash all your commits into one as this is a single feature and as such, i'd like to have it as a single commit.

Also, I had to upgrade a bunch of dependencies and java version so you'll need to rebase. Shouldn't cause any code conflicts but you'll get one conflict with the main pom.xml.

Thanks for doing all of this! Very much appreciated and yell at me if I don't respond fast enough!

File: pkts-streams/src/main/java/io/pkts/streams/impl/DefaultTcpStream.java [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+        // if new syn exchange, a new stream will be started by the synListener
+        // in that case, no need to add the new syn packet to the stream
+        if (fsm.getState() != TcpState.CLOSED_PORTS_REUSED){
+            packets.add(packet);
+        }
+    }
+
+    @Override
+    public TcpState getState(){
+        return (TcpState) fsm.getState();
```
[the remark this answers is not in the document]
Comment (Sebitosh) [reply pkts-streams/src/main/java/io/pkts/streams/impl/DefaultTcpStream.java:116]: I get this error when removing the cast:

`java: incompatible types: java.lang.Enum cannot be converted to io.pkts.streams.impl.tcpFSM.TcpStreamFSM.TcpState`

I'll put an issue on Hektor when I am done on all changes

File: pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java (hunk 4 above)
Comment (Sebitosh) [reply pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java:60]: Looks like it does guard it! I removed the condition

File: pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java (hunk 5 above)
Comment (Sebitosh) [reply pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java:194]: Took this as is from the `processFrame` functions in the Rtp and Sip handlers.

Should I use something like `logger.error("Exception thrown when processing TCP frame", e);` instead ?

File: pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java (hunk 6 above)
Comment (Sebitosh) [reply pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java:189]: They call on the "customer provided" tcpListener:

~~~java
    private void notifyStartStream(final TcpStream stream, final TCPPacket pkt) {
        if (this.tcpListener != null) {
            this.tcpListener.startStream(stream, pkt);
        }
    }

    private void notifyPacketReceived(final TcpStream stream, final TCPPacket pkt) {
        if (this.tcpListener != null) {
            this.tcpListener.packetReceived(stream, pkt);
        }
    }

    private void notifyEndStream(final TcpStream stream) {
        if (this.tcpListener != null) {
            this.tcpListener.endStream(stream);
        }
    }
~~~

I'll guard them like this:

~~~java
    private void notifyStartStream(final TcpStream stream, final TCPPacket pkt) {
     try {
        if (this.tcpListener != null) {
            this.tcpListener.startStream(stream, pkt);
        }
    } catch(Exception e) {
       logger.debug(Exception thrown from tcp stream listener, e)
   }
}
~~~

Also, took this code from the `SipStreamHandler`

File: pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java (hunk 7 above)
Comment (Sebitosh) [reply pkts-streams/src/main/java/io/pkts/streams/impl/TcpStreamHandler.java:205]: I see the point totally! To be honest this was my first project using more robust logging in java, and as such I have little notion of what is expected from a logging mechanism and what should be used where. Changed it to debug.

I took the code for this function from the `DefaultStreamHandler` .

File: pkts-streams/pom.xml [hunk 11]
```diff
@@ -8,7 +8,19 @@
   </parent>
 
   <name>Pkts Streams</name>
-  <groupId>io.pkts</groupId>
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
```
Comment (jonbo372) [inline pkts-streams/pom.xml:15]: you don't need this as I upgraded to 17. Also, you don't want to force different versions within the same package. https://github.com/aboutsip/pkts/blob/master/pom.xml#L193

**Review by jonbo372 — CHANGES_REQUESTED** [review]
Just one small thing and then let's merge and release!

**Review by jonbo372 — APPROVED** [review]
